### PR TITLE
linux: check CONFIG_CC_VERSION_TEXT matches our compiler

### PR DESCRIPTION
**Pull request by lrusak**
This seems to be an ongoing problem. I've created a simple way to adjust the kernel config using the script `./tools/adjust_kernel_config`. This script should always be used when adjusting the kernel configs. This sets up the build using the correct ARCH and toolchain that would be used to build the kernel.

I've included a sample diff of the RPi4 kernel config that shows the changes simply by using the script. I'm not sure how this even works and doesn't trip up `oldconfig` that is run during the build process.

sample output:
```shell
'gcc (Debian 10.2.1-6) 10.2.1 20210110' != 'aarch64-none-elf-gcc-11.2.0 (GCC) 11.2.0'
===================================================================================================
Please use the provided tools when adjusting the kernel config
For example: PROJECT=RPi ARCH=arm DEVICE=RPi4 ./tools/adjust_kernel_config
===================================================================================================
WARNING: CONFIG_CC_VERSION_TEXT doesn't match compiler version

'gcc' != 'aarch64-none-elf-gcc'
===================================================================================================
Please use the provided tools when adjusting the kernel config
For example: PROJECT=RPi ARCH=arm DEVICE=RPi4 ./tools/adjust_kernel_config
===================================================================================================
FAILURE: CONFIG_CC_VERSION_TEXT doesn't contain aarch64-none-elf-gcc
FAILURE: ./scripts/build linux during pre_make_target (package.mk)
```

The logic is that if the compiler version doesn't match just throw a warning but continue building (in case our compiler gets bumped and doesn't match).
If the compiler is different completely then we actually fail the build.

This will likely cause many of the builds to fail and the kernel configs for many projects should be updated. With this PR.

### CODE REVIEW

**Comment by HiassofT** [comment]
As discussed multiple times before breaking the build for no good reason is a really stupid idea.

Keep in mind that we may need to regression test compilers and switch to different gccs - like just recently where the linaro toolchain was broken. Being unable to build an image would be a big thumbs down then.

Printing a warning message is fine, but IMO  it's better to put this into tools/check_kernel_config where people will actually notice the warning

**Comment by lrusak** [comment]
> As discussed multiple times before breaking the build for no good reason is a really stupid idea.
> 

I'm not sure why you think there is no good reason. Having a consistent kernel config is the idea here. I don't want the options to change based on the compiler that is used. I want a consistent config every time. That was why I originally added the adjust_kernel_config script and the kernel_options file.

> 
> Keep in mind that we may need to regression test compilers and switch to different gccs - like just recently where the linaro toolchain was broken. Being unable to build an image would be a big thumbs down then.
> 

Eh? The build will fail if the compiler isn't the compiler that we use to build. If the version doesn't match only a warning is thrown. Switching the actual compiler would require updating the kernel config, again to create consistency.

> 
> Printing a warning message is fine, but IMO  it's better to put this into tools/check_kernel_config where people will actually notice the warning

I disagree. I want people to see that the build failed, why it failed, and what they need to do to fix it. Just shrugging it off doesn't help anyone.

**Comment by HiassofT** [comment]
We discussed it before and it's very simple, separation of concerns:

The build scripts are responsible for building, not testing or QA checks. If you want to test/check stuff that's a separate thing and usually placed in separate steps of your CI  pipeline.

You don't want to fail your build early because you then might miss further more important build problem (like kernel or modules actually not building).

CI can decide to fail your release pipeline and prevent publishing an official release if some checks are not met, but in no case should this fail the actual build  or upload of build artifacts- 101 of CI and testing.

Especially in this case, the compiler version used when creating the kconfig has almost no impact on the actual build. The kconfig diff for RPi contains mainly auto-generated settings that will be set by make oldconfig anyways. So this is a rather bad check as it'll cause too many false positives.

And, yes, we do want to be able to test eg with linaro or other toolchains in the future in case our compiler doesn't work correctly  - just as we had now the other way round. So compiler names can change and we don't want to go through the effort of changing 10+ kconfigs to actually regression test this just because an overly pedantic check prevented to build a kernel.

**Comment by lrusak** [comment]
> We discussed it before and it's very simple, separation of concerns:
> 
> 
> 
> The build scripts are responsible for building, not testing or QA checks. If you want to test/check stuff that's a separate thing and usually placed in separate steps of your CI  pipeline.
> 

This requires a few things. An actual CI pipeline to be setup correctly (which we don't have) and for someone to actively monitor the CI for warning, QA, etc. we don't have either of those things so it's better to catch them before they are even checked in.


> 
> You don't want to fail your build early because you then might miss further more important build problem (like kernel or modules actually not building).
> 

Failing is failing, you are going to miss other things regardless of when you fail.

We already miss a ton of QA checks because of safe_remove constantly throwing warnings.

> 
> CI can decide to fail your release pipeline and prevent publishing an official release if some checks are not met, but in no case should this fail the actual build  or upload of build artifacts- 101 of CI and testing.
> 

Again, it's just a pipe dream as we don't have that. You can save yourself the hassle by just committing correctly in the first place.

> 
> Especially in this case, the compiler version used when creating the kconfig has almost no impact on the actual build. The kconfig diff for RPi contains mainly auto-generated settings that will be set by make oldconfig anyways. So this is a rather bad check as it'll cause too many false positives.
> 

We should be using the tools we have to provide the best config possible. We should also be checking into source control as close as possible to what we are building.

oldconfig requires user interaction, so you might as well fail the build anyways.

> 
> And, yes, we do want to be able to test eg with linaro or other toolchains in the future in case our compiler doesn't work correctly  - just as we had now the other way round. So compiler names can change and we don't want to go through the effort of changing 10+ kconfigs to actually regression test this just because an overly pedantic check prevented to build a kernel.

I don't understand this paragraph. The compiler we use is the one that builds the kernel, so if you change the compiler you should change all relevant things for that.

This is why we have version control. It lets us check in exactly what we want.

**Comment by HiassofT** [comment]
Sigh. @CvH already stared redoing our CI mess so this is no excuse to add things to the wrong place just because this isn't finished yet.

If you want these checks so desperately in regular builds then help him get this going to speed things up.

You can just PR your checks to check_kernel_config (or some other script) now so everyone can use them and we can later decide if they are good to signal CI warnings or even fail the official release pipeline.

One thing for sure: you never ever should fail a development build with such overly pedantic checks - that's against the basic principles of making everything as robust as possible.

Adding check scripts and warnings to CI helps us as we can improve LE, but planting booby traps that trigger during development doesn't, this just annoys everyone.

**Comment by lrusak** [comment]
> Sigh. @CvH already stared redoing our CI mess so this is no excuse to add things to the wrong place just because this isn't finished yet.
> 
> If you want these checks so desperately in regular builds then help him get this going to speed things up.

So you think that blocking a PR based on some future possibility is a good idea? This corrects behavior now.

> You can just PR your checks to check_kernel_config (or some other script) now so everyone can use them and we can later decide if they are good to signal CI warnings or even fail the official release pipeline.
> 
> One thing for sure: you never ever should fail a development build with such overly pedantic checks - that's against the basic principles of making everything as robust as possible.

Why not? Yocto fails for many more things. The point is that if something can be corrected before even being submitted it should be. It's hardly being pedantic.

> Adding check scripts and warnings to CI helps us as we can improve LE, but planting booby traps that trigger during development doesn't, this just annoys everyone.

It's not a booby trap when it tells you how to correct it. CI isn't going to fix this as it's likely the CI won't be able to run for PR's anyway so we will always find out later.

**Comment by CvH** [comment]
instead of hard failing or just a warning message why not
`found error, press any key to ignore and build anyway`
would match both requirements I think ?

**Comment by HiassofT** [comment]
Which actual issue are you trying to solve with this PR?

I just ran  a test with and without this PR and the RPi4 .config used to build the kernel were 100% identical: http://ix.io/3X77

With this PR I get additional warnings though as you didn't update your RPi4 kconfig to the gcc used in LE master:
```
'aarch64-none-elf-gcc-11.2.0 (GCC) 11.2.0' != 'aarch64-none-elf-gcc-11.3.0 (GCC) 11.3.0'
===================================================================================================
Please use the provided tools when adjusting the kernel config
For example: PROJECT=RPi ARCH=arm DEVICE=RPi4 ./tools/adjust_kernel_config
===================================================================================================
WARNING: CONFIG_CC_VERSION_TEXT doesn't match compiler version
```

None of that has any impact on the build though so this PR isn't fixing an  actual issue, it creates a new one.

I'm not saying there isn't room for improvement in our config handling but breaking builds is the wrong approach.

**Comment by lrusak** [comment]
> Which actual issue are you trying to solve with this PR?
> 
> 
> 
> I just ran  a test with and without this PR and the RPi4 .config used to build the kernel were 100% identical: http://ix.io/3X77
> 

So why isn't that what is checked in? Of course this is difficult because we change some kernel config options based on distro options but still.


> 
> With this PR I get additional warnings though as you didn't update your RPi4 kconfig to the gcc used in LE master:
> 
> ```
> 
> 'aarch64-none-elf-gcc-11.2.0 (GCC) 11.2.0' != 'aarch64-none-elf-gcc-11.3.0 (GCC) 11.3.0'
> 
> ===================================================================================================
> 
> Please use the provided tools when adjusting the kernel config
> 
> For example: PROJECT=RPi ARCH=arm DEVICE=RPi4 ./tools/adjust_kernel_config
> 
> ===================================================================================================
> 
> WARNING: CONFIG_CC_VERSION_TEXT doesn't match compiler version
> 
> ```
> 
> 
> 
> None of that has any impact on the build though so this PR isn't fixing an  actual issue, it creates a new one.
> 

Good, it looks like it's working correctly then.

> 
> I'm not saying there isn't room for improvement in our config handling but breaking builds is the wrong approach.

Breaking builds because of bad configuration is completely valid IMO.

So we could fix it without failing, but then it will just get broken again down the line. Currently no one seems to care about these warnings but me.

**Comment by HiassofT** [comment]
The reason why it's not checked in is because
 1) it doesn't matter and
 2) I'm not a fan of committing transient information generated during kernell build into the repository, especially in case of 1) - it just creates clutter and merge conflicts for no good reason
3) creating the transient information requires at least a partial LE build (kernel toolchain)

I prefer a different workflow:

by just running ARCH=something make oldconfig I can avoid that clutter as the compiler on my machine doesn't change every couple of weeks, like in LE. So the diff in the commit reflects actual changes, can be easier reviewed and most of the time cleanly picked into LE10 as well (except the kernel version info comment at the beginning - there's some merit in having that so I commit it).

I do this automated at every kernel bump, review changes and commit them if anything exceprt the kernel info changed - so the actual kernel configs (minus transient info) is always up to date.

Every now and then I review that the config used to build the kernel is actually the same as when running oldconfig with the LE kernel toolchain - and as you see, they are.

I run this in a separate worktree / branch, where I don't build, building is done in my integration worktree / branch.

So, to summarize: you created a check that triggers a false positive (it assumes there's an issue when there actually is none) - this is not great at first. Then you think it should break the build because your false-positive check triggered. That's even worse for several reasons already outlined and is what I call planting booby traps.

And as I wrote in my very first comment we may want to warn about that, then this might become helpful in the future when the check is improved and doesn't produce false positives but warns about actual errors

**Comment by lrusak** [comment]
> The reason why it's not checked in is because
> 
> 1. it doesn't matter and
> 2. I'm not a fan of committing transient information generated during kernell build into the repository, especially in case of 1) - it just creates clutter and merge conflicts for no good reason
> 3. creating the transient information requires at least a partial LE build (kernel toolchain)
> 
> I prefer a different workflow:
> 
> by just running ARCH=something make oldconfig I can avoid that clutter as the compiler on my machine doesn't change every couple of weeks, like in LE. So the diff in the commit reflects actual changes, can be easier reviewed and most of the time cleanly picked into LE10 as well (except the kernel version info comment at the beginning - there's some merit in having that so I commit it).
> 
> I do this automated at every kernel bump, review changes and commit them if anything exceprt the kernel info changed - so the actual kernel configs (minus transient info) is always up to date.
> 
> Every now and then I review that the config used to build the kernel is actually the same as when running oldconfig with the LE kernel toolchain - and as you see, they are.
> 
> I run this in a separate worktree / branch, where I don't build, building is done in my integration worktree / branch.

Ok so why does your workflow trump everyone elses? We have the tool in the repo now and people should us it as the "official" way to make kernel config changes. What happens when I (or anyone else for that matter) makes changes using the script, PR's them, and they change things that you say is just "clutter"? Are you going to reject it? Are you going to ask them to change it? What if they are a first time contributor that isn't very good with git? What makes what you decide more important than anyone else?

If everyone just uses the script then the changes will be consistent.


> So, to summarize: you created a check that triggers a false positive (it assumes there's an issue when there actually is none) - this is not great at first. Then you think it should break the build because your false-positive check triggered. That's even worse for several reasons already outlined and is what I call planting booby traps.
> 
> And as I wrote in my very first comment we may want to warn about that, then this might become helpful in the future when the check is improved and doesn't produce false positives but warns about actual errors

So what you think is a booby trap is what I say is a configuration problem. Clearly there are changes in the kernel config that shows what should be included (even if when running oldconfig makes them equivalent (or close enough)).

Where do we go from here? Do we wait till someone else chimes in? Do we make everything only a warning and have all the warnings flood at the end of the build (which no one will fix most likely).

I would like to continue making changes to the `kernel_options` file like in #5219 which will basically make it required to use the `adjust_kernel_config` script so that we know everything is uniform. So this PR helps create an understanding of why a kernel configuration may be misconfigured.

**Comment by HiassofT** [comment]
It's quite simple: just commit the chunks with your actual changes - git add -p exists. This is what all contributors so far did and we never had any problem with that. You try to bring up a non-issue which is not helping because it has nothing to do with your wrong check and breaking builds.

Also please take a look at how other distributions set up their kernel configs and thing about why eg debian doesn't have any HAVE /  HAS settings include to their kernel config settings/fragments.

Right, because they are transient data which just happen to be stored in .config whenever kconfig is run - therefore there's no need to enforce rules on these trnasient settings - and I regularly check the very few setting dependent on these config values.

So, your warning check is completely wrong - the transient setting you check for will be adjusted a bit later anyway. If you want to check something there then look for absolutely important dependencies of the transient info. And at maximum warn about  that.

If you can't make the checks stable enough (avoiding false positives like your current check does) consider making these off by default and only enabled with an eg PEDANTIC option like c compilers, linters etc have - then this can actually be useful

**Comment by knaerzche** [comment]
@lrusak why don't you just check the GCC_OPTIONS *after* oldconfig run, i.e.  the config which will*really be taken for the actual build - like you are doing for all the other CONFIG-checks?
Those options will be replaced anyway. IMO we shouldn't have them in our (persistent) kernel-configs at all.

@HiassofT you really should check your tone. Your personal workflow isn't *that* important.

**Comment by HiassofT** [comment]
@knaerzche I'm really sorry about my tone but please understand that I'm extremely annoyed (and also very sad) to see that @lrusak ignored the result of the old warn vs die discussion which I thought we had already resolved a year ago (see #5044) and added "die" again to deliberately break builds.

Again: Warnings can be helpful, breaking builds is not (at least not without explicit opt-in via -Werror etc)

It's also not about my workflow, I merely added that as an explanation so you know how I'm managing RPi.

**Comment by vpeter4** [comment]
> Your personal workflow isn't _that_ important.

But workflow from the other is? ......

Seems over the the years LE forgot why it was forked from OE. New project name but same issues :(

**Comment by knaerzche** [comment]
> But workflow from the other is? ......

Well, I guess there not much builders building the LE kernel with a different compiler and using the config generated with this for building the LE. But sure: It should work for all scenarios.

> Seems over the the years LE forgot why it was forked from OE. New project name but same issues :(

Luckily I wasn't part of this, but I don't think so: Everybody is free to share his opinion/doubts - but I guess we always should stay friendly.
And this is just my personal opinion ;)

**Comment by stefansaraev** [comment]
in fact., CC_VERSION_TEXT does not matter at all. it's to make sure full kernel rebuild in kernel source tree if it wasn't cleaned up and compiler was updated. which again, doesnt matter for LE.
